sonic_installer/main.py

```diff
@@ -342,22 +344,25 @@ def migrate_sonic_packages(bootloader, binary_image_version):
             mount_bind(new_image_docker_dir, new_image_docker_mount)
             mount_procfs_chroot(new_image_mount)
             mount_sysfs_chroot(new_image_mount)
+            # Assume if docker.sh script exists we are installing Application Extension compatible image.
+            if not os.path.exists(os.path.join(new_image_mount, os.path.relpath(DOCKER_CTL_SCRIPT, os.path.abspath(os.sep)))):
+                echo_and_log("Warning: SONiC Application Extension is not supported in this image", LOG_WARN, fg="yellow")
+                return
             run_command_or_raise(["chroot", new_image_mount, DOCKER_CTL_SCRIPT, "start"])
+            docker_started = True
             run_command_or_raise(["cp", packages_path, os.path.join(new_image_mount, tmp_dir, packages_file)])
             run_command_or_raise(["touch", os.path.join(new_image_mount, "tmp", DOCKERD_SOCK)])
             run_command_or_raise(["mount", "--bind",
                                 os.path.join(VAR_RUN_PATH, DOCKERD_SOCK),
                                 os.path.join(new_image_mount, "tmp", DOCKERD_SOCK)])
             run_command_or_raise(["chroot", new_image_mount, "sh", "-c", "command -v {}".format(SONIC_PACKAGE_MANAGER)])
-        except SonicRuntimeException as err:
-            echo_and_log("Warning: SONiC Application Extension is not supported in this image: {}".format(err), LOG_ERR, fg="red")
-        else:
             run_command_or_raise(["chroot", new_image_mount, SONIC_PACKAGE_MANAGER, "migrate",
                                 os.path.join("/", tmp_dir, packages_file),
                                 "--dockerd-socket", os.path.join("/", tmp_dir, DOCKERD_SOCK),
                                 "-y"])
         finally:
-            run_command_or_raise(["chroot", new_image_mount, DOCKER_CTL_SCRIPT, "stop"], raise_exception=False)
+            if docker_started:
```

Comment:
>docker_started [](http://example.com/codeflow?start=15&length=14)

Why you need this check? If the DOCKER_CTL_SCRIPT could not be `stop` more than once, it's a bug.

Reply:
@qiluo-msft This check prevents doing "stop" in case we haven't done "start" at line 351

Reply:
I know your intention. My point here is that a service script in general should allow stop a stopped service and return 0. So I think this is a bug inside `/usr/lib/docker/docker.sh`

Reply:
It is not like it cannot stop it if it is stopped already. Here we added a check - if DOCKER_CTL_SCRIPT exists or not. If it exists then we continue the flow and start docker, if not we return from the function but a cleanup is still required. In case DOCKER_CTL_SCRIPT is missing this line will fail and print error. So a check is added "if docker_started", if it was started DOCKER_CTL_SCRIPT exists and we can execute this line, if not - not even needed to execute stop.

Reply:
> It is not like it cannot stop it if it is stopped already.

I checked `files/docker/docker`, and it is using `start-stop-daemon --stop --pidfile "$DOCKER_SSD_PIDFILE" --retry 10` in the stop branch. I think the option `--oknodo` will treat noop as success.

Reply:
@qiluo-msft It is not what I fix in this PR. The intention is to not call DOCKER_CTL_SCRIPT stop in case DOCKER_CTL_SCRIPT is missing.

Reply:
Thanks for the clarification!